proto/src/conversion.rs

```diff
@@ -0,0 +1,119 @@
+use crate::digest;
+use crate::error;
+use crate::tsmt;
+use miden_crypto::hash::rpo::RpoDigest;
+use miden_crypto::merkle::MerklePath;
+use miden_crypto::merkle::TieredSmtProof;
+use miden_crypto::Felt;
+use miden_crypto::FieldElement;
+use miden_crypto::StarkField;
+use miden_crypto::Word;
+
+impl From<[u64; 4]> for digest::Digest {
+    fn from(value: [u64; 4]) -> Self {
+        Self {
+            d0: value[0],
+            d1: value[1],
+            d2: value[2],
+            d3: value[3],
+        }
+    }
+}
+
+impl From<[Felt; 4]> for digest::Digest {
+    fn from(value: [Felt; 4]) -> Self {
+        Self {
+            d0: value[0].as_int(),
+            d1: value[1].as_int(),
+            d2: value[2].as_int(),
+            d3: value[3].as_int(),
+        }
+    }
+}
+
+impl From<RpoDigest> for digest::Digest {
+    fn from(value: RpoDigest) -> Self {
+        Self {
+            d0: value[0].as_int(),
+            d1: value[1].as_int(),
+            d2: value[2].as_int(),
+            d3: value[3].as_int(),
+        }
+    }
+}
+
+impl From<digest::Digest> for [u64; 4] {
+    fn from(value: digest::Digest) -> Self {
+        [value.d0, value.d1, value.d2, value.d3]
+    }
+}
+
+impl TryFrom<tsmt::NullifierProof> for TieredSmtProof {
+    type Error = error::ParseError;
+
+    fn try_from(value: tsmt::NullifierProof) -> Result<Self, Self::Error> {
+        let path = MerklePath::new(
+            value
+                .merkle_path
+                .into_iter()
+                .map(|v| v.try_into())
+                .collect::<Result<_, Self::Error>>()?,
+        );
+        let entries = value
+            .leaves
+            .into_iter()
+            .map(|leaf| {
+                let key = leaf.key.ok_or(error::ParseError::MissingLeafKey)?.try_into()?;
+                let value = [Felt::ZERO, Felt::ZERO, Felt::ZERO, Felt::new(leaf.value)];
+                let result = (key, value);
```

Comment:
We should probably document this somewhere (not sure where though) - i.e., the fact that the block number is stored in the last element of the leaf.